File: NeetCode/Trees/Binary_Tree_Maximum_Path_Sum/binary__tree__maximum__path__sum.py
from typing import Optional, List

from common.treenode import TreeNode, deserialize, printTree, null


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        maxFound = float("-inf")

        def dfs(node):
            nonlocal maxFound
            if node is None:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            allowToSplit = left + right + node.val
            maxChild = max(left, right, 0)
            maxFound = max(maxFound, allowToSplit, node.val, node.val + maxChild)
            return node.val + maxChild

        dfs(root)

        return maxFound


if __name__ == "__main__":
    sol = Solution()
    # root = [-15, 10, 20, null, null, 15, 5, -5]
    root = [2, -1]
    root = deserialize(root)

    print(sol.maxPathSum(root))
    print("Running Solution...")

File: NeetCode/Trees/Binary_Tree_Maximum_Path_Sum/Binary_Tree_Maximum_Path_Sum.cpp
// Binary_Tree_Maximum_Path_Sum.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    int maxPathSum(TreeNode *root)
    {
        int maxFound = INT_MIN;
        dfs(root, maxFound);
        return maxFound;
    }

    int dfs(TreeNode *root, int &maxFound)
    {
        if (!root)
        {
            return 0;
        }

        int left = dfs(root->left, maxFound);
        int right = dfs(root->right, maxFound);

        int maxChild = max(left, max(right, 0));
        int nPlusChild = root->val + maxChild;
        int split = left + right + root->val;
        maxFound = max({maxFound, split, nPlusChild, root->val});
        return nPlusChild;
    }
};
